tests/unit/discovery/windows/conftest.py

```diff
@@ -0,0 +1,169 @@
+from collections import defaultdict
+from contextlib import contextmanager
+from pathlib import Path
+
+import pytest
+
+
+@pytest.fixture()
+def _mock_registry(mocker):
+    from virtualenv.discovery.windows.pep514 import winreg
+
+    loc, glob = {}, {}
+    mock_value_str = (Path(__file__).parent / "winreg-mock-values.py").read_text()
+    exec(mock_value_str, glob, loc)
+    enum_collect = loc["enum_collect"]
+    value_collect = loc["value_collect"]
+    key_open = loc["key_open"]
+    hive_open = loc["hive_open"]
+
+    def _e(key, at):
+        key_id = key.value if isinstance(key, Key) else key
+        result = enum_collect[key_id][at]
+        if isinstance(result, OSError):
+            raise result
+        return result
+
+    mocker.patch.object(winreg, "EnumKey", side_effect=_e)
+
+    def _v(key, value_name):
+        key_id = key.value if isinstance(key, Key) else key
+        result = value_collect[key_id][value_name]
+        if isinstance(result, OSError):
+            raise result
+        return result
+
+    mocker.patch.object(winreg, "QueryValueEx", side_effect=_v)
+
+    class Key:
+        def __init__(self, value):
+            self.value = value
+
+        def __enter__(self):
+            return self
+
+        def __exit__(self, exc_type, exc_val, exc_tb):  # noqa: U100
+            return None
+
+    @contextmanager
+    def _o(*args):
+        if len(args) == 2:
+            key, value = args
+            key_id = key.value if isinstance(key, Key) else key
+            result = Key(key_open[key_id][value])  # this needs to be something that can be with-ed, so let's wrap it
+        elif len(args) == 4:
+            result = hive_open[args]
+        else:
+            raise RuntimeError
+        value = result.value if isinstance(result, Key) else result
+        if isinstance(value, OSError):
+            raise value
+        yield result
+
+    mocker.patch.object(winreg, "OpenKeyEx", side_effect=_o)
+    mocker.patch("os.path.exists", return_value=True)
+
+
+@pytest.fixture()
+def _collect_winreg_access(mocker):
+    # noinspection PyUnresolvedReferences
+    from winreg import EnumKey, OpenKeyEx, QueryValueEx
+
+    from virtualenv.discovery.windows.pep514 import winreg
+
+    hive_open = {}
+    key_open = defaultdict(dict)
+
+    @contextmanager
+    def _c(*args):
+        res = None
+        key_id = id(args[0]) if len(args) == 2 else None
+        try:
+            with OpenKeyEx(*args) as c:
+                res = id(c)
+                yield c
+        except Exception as exception:
+            res = exception
+            raise exception
+        finally:
+            if len(args) == 4:
+                hive_open[args] = res
+            elif len(args) == 2:
+                key_open[key_id][args[1]] = res
+
+    enum_collect = defaultdict(list)
+
+    def _e(key, at):
```

Comment:
a meaningful name for this method please

Reply:
Done